File: services/executor-node/src/lib/runner.ts
import { spawn } from 'child_process';
import { writeFile, rm, mkdir } from 'fs/promises';
import { join } from 'path';
import { tmpdir } from 'os';
import type { TestCase, TestResult } from '@code-practice/shared-types';

export interface RunResult {
  testResults: TestResult[];
  totalExecutionTimeMs: number;
  peakMemoryUsageBytes: number;
}

interface SingleTestResult {
  passed: boolean;
  actualOutput: unknown;
  executionTimeMs: number;
  memoryUsageBytes: number;
  error?: string;
}

/**
 * Execute compiled JavaScript code against test cases
 */
export async function runCode(
  compiledCode: string,
  functionName: string,
  testCases: TestCase[],
  onTestResult: (result: TestResult) => void,
): Promise<RunResult> {
  const workDir = join(tmpdir(), `executor-node-${Date.now()}`);
  await mkdir(workDir, { recursive: true });

  const testResults: TestResult[] = [];
  let totalExecutionTimeMs = 0;
  let peakMemoryUsageBytes = 0;

  try {
    for (let i = 0; i < testCases.length; i++) {
      const testCase = testCases[i];
      const result = await runSingleTest(
        workDir,
        compiledCode,
        functionName,
        testCase,
        i,
      );

      const testResult: TestResult = {
        testIndex: i,
        passed: result.passed,
        input: testCase.input,
        expectedOutput: testCase.expectedOutput,
        actualOutput: result.actualOutput,
        error: result.error,
        executionTimeMs: result.executionTimeMs,
        memoryUsageBytes: result.memoryUsageBytes,
      };

      testResults.push(testResult);
      totalExecutionTimeMs += result.executionTimeMs;
      peakMemoryUsageBytes = Math.max(peakMemoryUsageBytes, result.memoryUsageBytes);
      
      // Emit live update
      onTestResult(testResult);
    }

    return { testResults, totalExecutionTimeMs, peakMemoryUsageBytes };
  } finally {
    // Cleanup
    await rm(workDir, { recursive: true, force: true });
  }
}

async function runSingleTest(
  workDir: string,
  compiledCode: string,
  functionName: string,
  testCase: TestCase,
  testIndex: number,
): Promise<SingleTestResult> {
  const testFile = join(workDir, `test_${testIndex}.mjs`);
  
  // Create a test harness that runs the user function and reports results
  const harness = `
${compiledCode}

const input = ${JSON.stringify(testCase.input)};
const expected = ${JSON.stringify(testCase.expectedOutput)};

const startTime = performance.now();
const startMemory = process.memoryUsage().heapUsed;

try {
  const args = Array.isArray(input) ? input : [input];
  const result = ${functionName}(...args);
  const endTime = performance.now();
  const endMemory = process.memoryUsage().heapUsed;
  
  console.log(JSON.stringify({
    success: true,
    result,
    executionTimeMs: endTime - startTime,
    memoryUsageBytes: Math.max(0, endMemory - startMemory),
  }));
} catch (error) {
  const endTime = performance.now();
  console.log(JSON.stringify({
    success: false,
    error: error.message || String(error),
    executionTimeMs: endTime - startTime,
    memoryUsageBytes: 0,
  }));
}
`;

  await writeFile(testFile, harness);

  return new Promise((resolve) => {
    const startTime = performance.now();
    let output = '';
    let errorOutput = '';

    const child = spawn('node', [testFile], {
      timeout: 10000, // 10 second timeout
      cwd: workDir,
    });

    child.stdout.on('data', (data) => {
      output += data.toString();
    });

    child.stderr.on('data', (data) => {
      errorOutput += data.toString();
    });

    child.on('close', (code) => {
      const wallTime = performance.now() - startTime;

      if (code !== 0 || errorOutput) {
        resolve({
          passed: false,
          actualOutput: undefined,
          executionTimeMs: wallTime,
          memoryUsageBytes: 0,
          error: errorOutput || `Process exited with code ${code}`,
        });
        return;
      }

      try {
        const result = JSON.parse(output.trim());
        
        if (!result.success) {
          resolve({
            passed: false,
            actualOutput: undefined,
            executionTimeMs: result.executionTimeMs,
            memoryUsageBytes: result.memoryUsageBytes,
            error: result.error,
          });
          return;
        }

        const passed = JSON.stringify(result.result) === JSON.stringify(testCase.expectedOutput);
        
        resolve({
          passed,
          actualOutput: result.result,
          executionTimeMs: result.executionTimeMs,
          memoryUsageBytes: result.memoryUsageBytes,
        });
      } catch {
        resolve({
          passed: false,
          actualOutput: output,
          executionTimeMs: wallTime,
          memoryUsageBytes: 0,
          error: 'Failed to parse test output',
        });
      }
    });

    child.on('error', (err) => {
      resolve({
        passed: false,
        actualOutput: undefined,
        executionTimeMs: performance.now() - startTime,
        memoryUsageBytes: 0,
        error: err.message,
      });
    });
  });
}

File: services/executor-python/runner.py
import asyncio
import json
import shutil
import sys
import tempfile
from dataclasses import dataclass
from pathlib import Path
from time import perf_counter, perf_counter_ns
from typing import Awaitable, Callable, Dict, List

from models import TestCase, TestResult


@dataclass
class CompileResult:
    success: bool
    code: str | None
    compileTimeMs: float
    error: str | None = None


@dataclass
class RunResult:
    testResults: List[TestResult]
    totalExecutionTimeMs: float
    peakMemoryUsageBytes: int


async def compile_python(code: str) -> CompileResult:
    """Perform a lightweight syntax check for Python code and measure time."""
    start_ns = perf_counter_ns()
    try:
        compile(code, "<user_code>", "exec")
    except SyntaxError as exc:
        elapsed_ms = (perf_counter_ns() - start_ns) / 1_000_000
        return CompileResult(False, None, elapsed_ms, f"{exc.__class__.__name__}: {exc}")

    elapsed_ms = (perf_counter_ns() - start_ns) / 1_000_000
    return CompileResult(True, code, elapsed_ms, None)


async def run_code(
    compiled_code: str,
    function_name: str,
    test_cases: List[TestCase],
    on_test_result: Callable[[TestResult], Awaitable[None]],
) -> RunResult:
    """Execute compiled Python code against all test cases in a temp workspace."""
    work_dir_path = Path(tempfile.mkdtemp(prefix="executor-python-"))
    test_results: List[TestResult] = []
    total_execution_ms = 0.0
    peak_memory_bytes = 0

    try:
        for index, test_case in enumerate(test_cases):
            single = await _run_single_test(
                work_dir_path, compiled_code, function_name, test_case, index
            )

            result = TestResult(
                testIndex=index,
                passed=single["passed"],
                input=test_case.input,
                expectedOutput=test_case.expectedOutput,
                actualOutput=single.get("actualOutput"),
                error=single.get("error"),
                executionTimeMs=single["executionTimeMs"],
                memoryUsageBytes=single["memoryUsageBytes"],
            )

            test_results.append(result)
            total_execution_ms += single["executionTimeMs"]
            peak_memory_bytes = max(peak_memory_bytes, single["memoryUsageBytes"])

            await on_test_result(result)
    finally:
        shutil.rmtree(work_dir_path, ignore_errors=True)

    return RunResult(test_results, total_execution_ms, peak_memory_bytes)


async def _run_single_test(
    work_dir: Path,
    compiled_code: str,
    function_name: str,
    test_case: TestCase,
    index: int,
) -> Dict[str, object]:
    """Run a single test case in an isolated subprocess."""
    script_path = work_dir / f"test_{index}.py"
    harness = _build_harness(compiled_code, function_name, test_case)
    script_path.write_text(harness, encoding="utf-8")

    start_wall = perf_counter()
    proc = await asyncio.create_subprocess_exec(
        sys.executable,
        str(script_path),
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
        cwd=str(work_dir),
    )

    stdout, stderr = await proc.communicate()
    wall_ms = (perf_counter() - start_wall) * 1000.0

    if proc.returncode != 0 or stderr:
        return {
            "passed": False,
            "actualOutput": None,
            "executionTimeMs": wall_ms,
            "memoryUsageBytes": 0,
            "error": stderr.decode() or f"Process exited with code {proc.returncode}",
        }

    output_lines = stdout.decode().strip().splitlines()
    if not output_lines:
        return {
            "passed": False,
            "actualOutput": None,
            "executionTimeMs": wall_ms,
            "memoryUsageBytes": 0,
            "error": "No output from test harness",
        }

    try:
        data = json.loads(output_lines[-1])
    except json.JSONDecodeError as exc:
        return {
            "passed": False,
            "actualOutput": None,
            "executionTimeMs": wall_ms,
            "memoryUsageBytes": 0,
            "error": f"Invalid JSON from harness: {exc}",
        }

    if not data.get("success"):
        return {
            "passed": False,
            "actualOutput": None,
            "executionTimeMs": float(data.get("executionTimeMs", wall_ms)),
            "memoryUsageBytes": int(data.get("memoryUsageBytes", 0)),
            "error": data.get("error") or "Unknown error in harness",
        }

    return {
        "passed": True,
        "actualOutput": data.get("result"),
        "executionTimeMs": float(data.get("executionTimeMs", wall_ms)),
        "memoryUsageBytes": int(data.get("memoryUsageBytes", 0)),
    }


def _build_harness(code: str, function_name: str, test_case: TestCase) -> str:
    """Build the Python harness script that wraps user code and a single test."""
    test_input = json.dumps(test_case.input)

    # The harness prints a single JSON line describing the outcome of the call.
    return f"""import json
import time
import tracemalloc

{code}

def _run():
    input_value = {test_input}
    args = input_value if isinstance(input_value, list) else [input_value]
    tracemalloc.start()
    start_ns = time.perf_counter_ns()
    try:
        result = {function_name}(*args)
        end_ns = time.perf_counter_ns()
        current, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        print(json.dumps({{
            "success": True,
            "result": result,
            "executionTimeMs": (end_ns - start_ns) / 1_000_000,
            "memoryUsageBytes": int(peak),
        }}, default=str))
    except Exception as exc:  # noqa: BLE001
        end_ns = time.perf_counter_ns()
        tracemalloc.stop()
        print(json.dumps({{
            "success": False,
            "error": str(exc),
            "executionTimeMs": (end_ns - start_ns) / 1_000_000,
            "memoryUsageBytes": 0,
        }}, default=str))


if __name__ == "__main__":
    _run()
"""

